In [146]:
import cv2
import numpy as np

# --------------------------------------------------------------------------------------------------------------

# Image Thresholding

## Thresholding is one of the most common (and basic) segmentation techniques in computer vision and it allows us to separate the foreground (i.e., the objects that we are interested in) from the background of the image.

## Thresholding is mainly of Three Types:-

### 1)Simple Thresholding
### 2)OTSU Thresholding
### 3)Adaptive Thresholding

## ---------------------------------------------------------------------------------------------------------

## -->Simple Thresholding

### Its also termed as BINARY Thresholding.
### In this Technique we Manually set a Standard Threshold Value(T) and each Pixel of the Image is compared with the Threshold Value(T).
### -->If the Pixel Value is Less than T then that Pixel Value is set to 0 in the Output Image.
### -->If the Pixel Value is Greater than T then that Pixel Value is set to the Maximum Value(Provided in the Method Argument as THIRD Argument) OR Some Value depending on the THRESHOLDING TYPE.
### It works very well in a Controlled Lighting Conditions where we can Ensure High Contrast between Background and Foreground.

### To perform Simple Threshold we use `cv2.threshold()` Method.

## For `cv2.threshold()` :-

### < First Argument >:-It takes the NUMPY Array of the Input Image.
### < Second Argument >:-It takes the Threshold Value(T).
### < Third Argument >:-It takes the Maximum Value(Only Used for THRESH_BINARY or THRESH_BINARY_INV) .
### < Fourth Argument >:-It takes the THRESHOLDING TYPE Flags. 

#### cv2.threshold():-[LINK](https://docs.opencv.org/4.x/d7/d1b/group__imgproc__misc.html#gae8a4a146d1ca78c626a53577199e9c57)

#### THRESHOLDING Type FLAGS:-[LINK](https://docs.opencv.org/4.x/d7/d1b/group__imgproc__misc.html#gaa9e58d2860d4afa658ef70a9b1115576)

## Simpe Thresholding can also be sub-divided according to the THRESHOLDING TYPE Flag being used:-

### ------------------------------------------------------------------------------------------

### Thresh Binary

### In this each Pixel Value of the Input Image is compared to the Threshold Value(T) And If:-
### -->The Pixel Value is Greater than T then it Outputs to the Maximum Value(Provided in the Third Argument).
### -->The Pixel Value is Less than or Equal to T then it Outputs to 0.

In [12]:
cn1color = cv2.imread(r"E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\InputData\coins1.png")
cn1gray = cv2.cvtColor(cn1color,cv2.COLOR_BGR2GRAY)

In [4]:
cv2.imshow("Colored Image",cn1color)
cv2.imshow("Graye Scale Image",cn1gray)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [43]:
TB1gray = cv2.threshold(cn1gray,180,255,cv2.THRESH_BINARY)

In [11]:
TB1gray

(180.0,
 array([[255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        ...,
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255]], dtype=uint8))

### We can see It Returns a Tuple containing:-
### -->Threshold Value as First Argument.
### -->Thresh Binary Image as Second Argument.

In [7]:
cv2.imshow("Gray Scale Image",cn1gray)
cv2.imshow("Threshold Binary GrayScale Image",TB1gray[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see we Performed THRESH Binary Simple Thresholding(with T =180) on Gray Scale Image and got the Expected Results.

In [44]:
TB1color = cv2.threshold(cn1color,180,255,cv2.THRESH_BINARY)

In [17]:
TB1color

(180.0,
 array([[[255, 255, 255],
         [255, 255, 255],
         [255, 255, 255],
         ...,
         [255, 255, 255],
         [255, 255, 255],
         [255, 255, 255]],
 
        [[255, 255, 255],
         [255, 255, 255],
         [255, 255, 255],
         ...,
         [255, 255, 255],
         [255, 255, 255],
         [255, 255, 255]],
 
        [[255, 255, 255],
         [255, 255, 255],
         [255, 255, 255],
         ...,
         [255, 255, 255],
         [255, 255, 255],
         [255, 255, 255]],
 
        ...,
 
        [[255, 255, 255],
         [255, 255, 255],
         [255, 255, 255],
         ...,
         [255, 255, 255],
         [255, 255, 255],
         [255, 255, 255]],
 
        [[255, 255, 255],
         [255, 255, 255],
         [255, 255, 255],
         ...,
         [255, 255, 255],
         [255, 255, 255],
         [255, 255, 255]],
 
        [[255, 255, 255],
         [255, 255, 255],
         [255, 255, 255],
         ...,
         [255, 255, 

In [18]:
cv2.imshow("Colored Image",cn1color)
cv2.imshow("Threshold Binary Colored Image",TB1color[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see we Performed Thresh Binary Simple Thresholding(with T=180) on an Color Image and got an Unexpected Result.

### As we can see we got Better Results when Performed Thresholding on Gray Scale Image So Performing Following Tasks.

In [20]:
TB1_240 = cv2.threshold(cn1gray,240,255,cv2.THRESH_BINARY)
TB1_220 = cv2.threshold(cn1gray,220,255,cv2.THRESH_BINARY)
TB1_200 = cv2.threshold(cn1gray,200,255,cv2.THRESH_BINARY)
TB1_180 = cv2.threshold(cn1gray,180,255,cv2.THRESH_BINARY)
TB1_150= cv2.threshold(cn1gray,150,255,cv2.THRESH_BINARY)
TB1_130 = cv2.threshold(cn1gray,130,255,cv2.THRESH_BINARY)
TB1_110 = cv2.threshold(cn1gray,110,255,cv2.THRESH_BINARY)
TB1_90 = cv2.threshold(cn1gray,90,255,cv2.THRESH_BINARY)
TB1_50 = cv2.threshold(cn1gray,50,255,cv2.THRESH_BINARY)
TB1_30 = cv2.threshold(cn1gray,30,255,cv2.THRESH_BINARY)
TB1_10 = cv2.threshold(cn1gray,10,255,cv2.THRESH_BINARY)

In [21]:
cv2.imshow("Normal Image",cn1gray)
cv2.imshow("fg240",TB1_240[1])
cv2.imshow("fg220",TB1_220[1])
cv2.imshow("fg200",TB1_200[1])
cv2.imshow("fg180",TB1_180[1])
cv2.imshow("fg150",TB1_150[1])
cv2.imshow("fg110",TB1_110[1])
cv2.imshow("fg90",TB1_90[1])
cv2.imshow("fg50",TB1_50[1])
cv2.imshow("fg30",TB1_30[1])
cv2.imshow("fg10",TB1_10[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see how Different Threshold Value Affects the Output Image for Simple Binary Thresholding on Gray Scale Image.

### ------------------------------------------------------------------------------------------

## Thresh Binary Inverse

### In this each Pixel Value of the Image Is Compared to the Threshold Value(T) And If:-
### -->The Pixel Value is Greater than T then it output to 0.
### -->The Pixel Value is Less than T then it Outputs to the Maximum Value.(Provided in the Third Argument).

### **Note:-Its the opposite of the THRESH Binary

In [45]:
TBI1_gray = cv2.threshold(cn1gray,180,255,cv2.THRESH_BINARY_INV)

In [6]:
cv2.imshow("Normal GrayScale Image",cn1gray)
cv2.imshow("Threshold Binary Inverse",TBI1_gray[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We performed Simple Binary Inverse Threshold(with T =180) on a Gray Scale Image and Obtained the result as expected

In [46]:
TBI1_color = cv2.threshold(cn1color,180,255,cv2.THRESH_BINARY_INV)

In [9]:
cv2.imshow("Normal Colored Image",cn1color)
cv2.imshow("Threshold Binary Inverse Colored Image",TBI1_color[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We performed Simple Binary Inverse Threshold(with T=180) on Colored Image and Obtained Unexpected Results

### As we can see we got Better and Expected Results on performing Simple Binary Inverse Thresholding hence Performing the following Tasks

In [10]:
TBI1_250 = cv2.threshold(cn1gray,250,255,cv2.THRESH_BINARY_INV)
TBI1_225 = cv2.threshold(cn1gray,225,255,cv2.THRESH_BINARY_INV)
TBI1_200 = cv2.threshold(cn1gray,200,255,cv2.THRESH_BINARY_INV)
TBI1_180 = cv2.threshold(cn1gray,180,255,cv2.THRESH_BINARY_INV)
TBI1_160 = cv2.threshold(cn1gray,160,255,cv2.THRESH_BINARY_INV)
TBI1_140 = cv2.threshold(cn1gray,140,255,cv2.THRESH_BINARY_INV)
TBI1_120 = cv2.threshold(cn1gray,120,255,cv2.THRESH_BINARY_INV)
TBI1_100 = cv2.threshold(cn1gray,100,255,cv2.THRESH_BINARY_INV)
TBI1_80 = cv2.threshold(cn1gray,80,255,cv2.THRESH_BINARY_INV)
TBI1_50 = cv2.threshold(cn1gray,50,255,cv2.THRESH_BINARY_INV)
TBI1_30 = cv2.threshold(cn1gray,30,255,cv2.THRESH_BINARY_INV)
TBI1_10 = cv2.threshold(cn1gray,10,255,cv2.THRESH_BINARY_INV)

In [12]:
cv2.imshow("TBI1_250",TBI1_250[1])
cv2.imshow("TBI1_225",TBI1_225[1])
cv2.imshow("TBI1_200",TBI1_200[1])
cv2.imshow("TBI1_180",TBI1_180[1])
cv2.imshow("TBI1_160",TBI1_160[1])
cv2.imshow("TBI1_140",TBI1_140[1])
cv2.imshow("TBI1_120",TBI1_120[1])
cv2.imshow("TBI1_100",TBI1_100[1])
cv2.imshow("TBI1_80",TBI1_80[1])
cv2.imshow("TBI1_50",TBI1_50[1])
cv2.imshow("TBI1_30",TBI1_30[1])
cv2.imshow("TBI1_10",TBI1_10[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see in the Above Cells how Different Vaue of Threshold can affect the Results when performing Simple Binary Inverse Thresholding on Gray Scale Images

### ------------------------------------------------------------------------------------------

## Thresh Trunc

### In this each Pixel Value of the Image Is Compared to the Threshold Value(T) And If:-
### -->The Pixel Value is Greater than T then it output to T.
### -->The Pixel Value is Less than T then it doesnot Changes its Value.

In [14]:
TT1_gray = cv2.threshold(cn1gray,180,255,cv2.THRESH_TRUNC)

In [15]:
cv2.imshow("Gray Scale Image",cn1gray)
cv2.imshow("Thresh Trunc GrayScale Image",TT1_gray[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We performed Simple Trunc Thresholding(with T=180) on Gray Scale Image and Obtained Expected Result

In [16]:
TT1_color = cv2.threshold(cn1color,180,255,cv2.THRESH_TRUNC)

In [18]:
cv2.imshow("Colored Image",cn1color)
cv2.imshow("Thresh Trunc Colored Image",TT1_color[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We performed Simple Trunc Thresholding(with T=180) on Colored Image and Obtain UnExpected Result

### As we can see we got Better and Expected Result on GrayScale Images hence Performing the Following Tasks

In [20]:
TT1_250 = cv2.threshold(cn1gray,250,255,cv2.THRESH_TRUNC)
TT1_220 = cv2.threshold(cn1gray,220,255,cv2.THRESH_TRUNC)
TT1_200 = cv2.threshold(cn1gray,200,255,cv2.THRESH_TRUNC)
TT1_180 = cv2.threshold(cn1gray,180,255,cv2.THRESH_TRUNC)
TT1_160 = cv2.threshold(cn1gray,160,255,cv2.THRESH_TRUNC)
TT1_140 = cv2.threshold(cn1gray,140,255,cv2.THRESH_TRUNC)
TT1_120 = cv2.threshold(cn1gray,120,255,cv2.THRESH_TRUNC)
TT1_100 = cv2.threshold(cn1gray,100,255,cv2.THRESH_TRUNC)
TT1_80 = cv2.threshold(cn1gray,80,255,cv2.THRESH_TRUNC)
TT1_50 = cv2.threshold(cn1gray,50,255,cv2.THRESH_TRUNC)
TT1_30 = cv2.threshold(cn1gray,30,255,cv2.THRESH_TRUNC)
TT1_10 = cv2.threshold(cn1gray,10,255,cv2.THRESH_TRUNC)

In [23]:
cv2.imshow("TT1_250",TT1_250[1])
cv2.imshow("TT1_220",TT1_220[1])
cv2.imshow("TT1_200",TT1_200[1])
cv2.imshow("TT1_180",TT1_180[1])
cv2.imshow("TT1_160",TT1_160[1])
cv2.imshow("TT1_140",TT1_140[1])
cv2.imshow("TT1_120",TT1_120[1])
cv2.imshow("TT1_100",TT1_100[1])
cv2.imshow("TT1_80",TT1_80[1])
cv2.imshow("TT1_50",TT1_50[1])
cv2.imshow("TT1_30",TT1_30[1])
cv2.imshow("TT1_10",TT1_10[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see in the Above Cells that how Different Threshold Value while performing Simple Thrunc Thresholding can affect the results.

### ------------------------------------------------------------------------------------------

## Thresh To Zero

### In this each Pixel Value of the Image Is Compared to the Threshold Value(T) And If:-
### -->The Pixel Value is Greater than T then it doesnot Changes its Value.
### -->The Pixel Value is Less than T then it Outputs to 0.

In [24]:
TZ1_gray = cv2.threshold(cn1gray,180,255,cv2.THRESH_TOZERO)

In [27]:
cv2.imshow("Normal Gray Scale Image",cn1gray)
cv2.imshow("Thresh To Zero Gray Scale Image",TZ1_gray[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We performed Simple Thresh To Zero Thresholding(with T=180) on Gray Scale Image and got Expected Results

In [28]:
TZ1_color = cv2.threshold(cn1color,180,255,cv2.THRESH_TOZERO) 

In [29]:
cv2.imshow("Normal Colored Image",cn1color)
cv2.imshow("Thresh To Zero Colored Image",TZ1_color[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We performed Simple Thresh To Zerp Thresholding(with T=180) on Colored Image and got the UnExpected Results

### As we can see we got Better and Expected Results on GrayScale Images hence performing following Tasks

In [30]:
TZ1_250 = cv2.threshold(cn1gray,250,255,cv2.THRESH_TOZERO)
TZ1_220 = cv2.threshold(cn1gray,220,255,cv2.THRESH_TOZERO)
TZ1_200 = cv2.threshold(cn1gray,200,255,cv2.THRESH_TOZERO)
TZ1_180 = cv2.threshold(cn1gray,180,255,cv2.THRESH_TOZERO)
TZ1_160 = cv2.threshold(cn1gray,160,255,cv2.THRESH_TOZERO)
TZ1_140 = cv2.threshold(cn1gray,140,255,cv2.THRESH_TOZERO)
TZ1_120 = cv2.threshold(cn1gray,120,255,cv2.THRESH_TOZERO)
TZ1_100 = cv2.threshold(cn1gray,100,255,cv2.THRESH_TOZERO)
TZ1_80 = cv2.threshold(cn1gray,80,255,cv2.THRESH_TOZERO)
TZ1_60 = cv2.threshold(cn1gray,60,255,cv2.THRESH_TOZERO)
TZ1_50 = cv2.threshold(cn1gray,50,255,cv2.THRESH_TOZERO)
TZ1_30 = cv2.threshold(cn1gray,30,255,cv2.THRESH_TOZERO)
TZ1_10 = cv2.threshold(cn1gray,10,255,cv2.THRESH_TOZERO)

In [32]:
cv2.imshow("TZ1_250",TZ1_250[1])
cv2.imshow("TZ1_220",TZ1_220[1])
cv2.imshow("TZ1_200",TZ1_200[1])
cv2.imshow("TZ1_180",TZ1_180[1])
cv2.imshow("TZ1_160",TZ1_160[1])
cv2.imshow("TZ1_140",TZ1_140[1])
cv2.imshow("TZ1_120",TZ1_120[1])
cv2.imshow("TZ1_100",TZ1_100[1])
cv2.imshow("TZ1_80",TZ1_80[1])
cv2.imshow("TZ1_60",TZ1_60[1])
cv2.imshow("TZ1_50",TZ1_50[1])
cv2.imshow("TZ1_30",TZ1_30[1])
cv2.imshow("TZ1_10",TZ1_10[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see in Above Cells that how Different Threshold Value in Simple Tresh To Zero can affect Results

### ------------------------------------------------------------------------------------------

## Thresh To Zero Inverse

### In this each Pixel Value of the Image Is Compared to the Threshold Value(T) And If:-
### -->The Pixel Value is Greater than T then it Outputs to 0.
### -->The Pixel Value is Less than T then it doesnot Changes.

### **NOTE:-Its the opposite of THRESH To Zero Thresholding

In [33]:
TTZI1_gray = cv2.threshold(cn1gray,180,255,cv2.THRESH_TOZERO_INV)

In [34]:
cv2.imshow("Gray Scale Image",cn1gray)
cv2.imshow("Thresh To Zero Gray Scale Image",TTZI1_gray[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We have performed Simple Thresh To Zero Thresholding(with T=180) on a Gray Scale Image and got the Expected Results

In [36]:
TTZI1_color = cv2.threshold(cn1color,180,255,cv2.THRESH_TOZERO_INV)

In [37]:
cv2.imshow("Normal Colored Image",cn1color)
cv2.imshow("Thresh To Zero Colored Image",TTZI1_color[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We have performed Simple Thresh To Zero Thresholding(with T=180) on a Colored Image and got UnExpected Results

### As we can see that we got Better and Expected Results on Gray Scale Images hence performing the following tasks

In [40]:
TTZI_250 = cv2.threshold(cn1gray,250,255,cv2.THRESH_TOZERO_INV)
TTZI_240 = cv2.threshold(cn1gray,240,255,cv2.THRESH_TOZERO_INV)
TTZI_220 = cv2.threshold(cn1gray,220,255,cv2.THRESH_TOZERO_INV)
TTZI_200 = cv2.threshold(cn1gray,200,255,cv2.THRESH_TOZERO_INV)
TTZI_180 = cv2.threshold(cn1gray,180,255,cv2.THRESH_TOZERO_INV)
TTZI_160 = cv2.threshold(cn1gray,160,255,cv2.THRESH_TOZERO_INV)
TTZI_140 = cv2.threshold(cn1gray,140,255,cv2.THRESH_TOZERO_INV)
TTZI_120 = cv2.threshold(cn1gray,120,255,cv2.THRESH_TOZERO_INV)
TTZI_100 = cv2.threshold(cn1gray,100,255,cv2.THRESH_TOZERO_INV)
TTZI_80 = cv2.threshold(cn1gray,80,255,cv2.THRESH_TOZERO_INV)
TTZI_50 = cv2.threshold(cn1gray,50,255,cv2.THRESH_TOZERO_INV)
TTZI_30 = cv2.threshold(cn1gray,30,255,cv2.THRESH_TOZERO_INV)
TTZI_10 = cv2.threshold(cn1gray,10,255,cv2.THRESH_TOZERO_INV)

In [42]:
cv2.imshow("TTZI1_250",TTZI_250[1])
cv2.imshow("TTZI1_240",TTZI_240[1])
cv2.imshow("TTZI1_220",TTZI_220[1])
cv2.imshow("TTZI1_200",TTZI_200[1])
cv2.imshow("TTZI1_180",TTZI_180[1])
cv2.imshow("TTZI1_160",TTZI_160[1])
cv2.imshow("TTZI1_140",TTZI_140[1])
cv2.imshow("TTZI1_120",TTZI_120[1])
cv2.imshow("TTZI1_100",TTZI_100[1])
cv2.imshow("TTZI1_80",TTZI_80[1])
cv2.imshow("TTZI1_50",TTZI_50[1])
cv2.imshow("TTZI1_30",TTZI_30[1])
cv2.imshow("TTZI1_10",TTZI_10[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see that in Above Cells that how Different Threshold Value for Simple Thresh To Zero Inverse affects the Results

### ------------------------------------------------------------------------------------------

## Comparing The Results of all the Five Simple Thresholding Types(with T =180) on Gray Scale Image

In [48]:
cv2.imshow("Normal Gray Scale Image",cn1gray)
cv2.imshow("Thresh Binary",TB1gray[1])
cv2.imshow("Thresh Binary Inverse",TBI1_gray[1])
cv2.imshow("Thresh Trunc",TT1_gray[1])
cv2.imshow("Thresh To Zero",TZ1_gray[1])
cv2.imshow("Thresh To Zero Inverse",TTZI1_gray[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

## Comparing The Results of all the Five Simple Thresholding Types(with T =180) on Colored Image

In [49]:
cv2.imshow("Colored Image",cn1color)
cv2.imshow("Thresh Binary",TB1color[1])
cv2.imshow("Thresh Binary Inverse",TBI1_color[1])
cv2.imshow("Thresh Trunc",TT1_color[1])
cv2.imshow("Thresh To Zero",TZ1_color[1])
cv2.imshow("Thresh To Zero Inverse",TTZI1_color[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

## ---------------------------------------------------------------------------------------------------------

## OTSU Thresholding

### As we know in Simple Thresholding we manually provide the Threshold Value(T) and this works well for Images in Controlled Lightning Conditions.

### But in Real-world Images where we have no prior Knowledge of the Environment and Lightning Conditions, there we can Automatically compute an Threshold Value using OTSU Thresholding.

### Hence ,OTSU Thresholding is an Image Thresholding Technique which assumes that the Gray Scale Histogram of our Image Pixel Intensties is Bi-Modal Or In short means that our image contains only two Classes of Pixels BACKGROUND and FOREGROUND.

### Based on the GrayScale Histogram,OTSU Threshold Technique calculates the Optimal Threshold Value(T) such that the Variance between BackGround and ForeGround is Minimal.

In [15]:
cn1_TBO = cv2.threshold(cn1gray,0,255,cv2.THRESH_BINARY | cv2.THRESH_OTSU)

### We can see that we just Passed Threshold Value(T in Second Argument) as 0 but it will not be used but will be calculated by the OTSU Technique

In [112]:
cn1_TBO

(185.0,
 array([[255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        ...,
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255]], dtype=uint8))

In [114]:
cv2.imshow("GrayScale Image",cn1gray)
cv2.imshow("Thresh Binary OTSU",cn1_TBO[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see in the Above Cells that we have Performed Binary Otsu Thresholding on Gray Scale Image,and we can see that OTSU Technique calculated The Optimal Threshold Value(T) as 185,and we can also notice that the Reslus are Good and Expected

In [115]:
cn1_TBIO = cv2.threshold(cn1gray,0,255,cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)
cn1_TTO = cv2.threshold(cn1gray,0,255,cv2.THRESH_TRUNC | cv2.THRESH_OTSU)
cn1_TTZO = cv2.threshold(cn1gray,0,255,cv2.THRESH_TOZERO | cv2.THRESH_OTSU)
cn1_TTZIO = cv2.threshold(cn1gray,0,255,cv2.THRESH_TOZERO_INV | cv2.THRESH_OTSU)

In [116]:
cn1_TBIO

(185.0,
 array([[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=uint8))

In [117]:
cn1_TTO

(185.0,
 array([[185, 185, 185, ..., 185, 185, 185],
        [185, 185, 185, ..., 185, 185, 185],
        [185, 185, 185, ..., 185, 185, 185],
        ...,
        [185, 185, 185, ..., 185, 185, 185],
        [185, 185, 185, ..., 185, 185, 185],
        [185, 185, 185, ..., 185, 185, 185]], dtype=uint8))

In [118]:
cn1_TTZO

(185.0,
 array([[255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        ...,
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255]], dtype=uint8))

In [119]:
cn1_TTZIO

(185.0,
 array([[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=uint8))

In [121]:
cv2.imshow("Normal Gray Scale Image",cn1gray)
cv2.imshow("Thresh Binary OTSU",cn1_TBO[1])
cv2.imshow("Thresh Binary Inverse OTSU",cn1_TBIO[1])
cv2.imshow("Thresh Trunc OTSU",cn1_TTO[1])
cv2.imshow("Thresh To Zero OTSU",cn1_TTZO[1])
cv2.imshow("Thresh To Zero Inverse OTSU",cn1_TTZIO[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see that in above cells we perform all types of Simple Thresholding with OTSU Technique and we can notice that:-

### -->That OTSU calculates a Single Threshold Value(T) irrespective of the Simple Threshold Technique being used.

## .)Limitations of OTSU Threshold and Global Thresholding:-

### -->`It Doesnot Works Properly with Images that have Noise`:-If Noise is Present in the Image then it reduces the Sharp Valleys between the peaks of the Bi-Modal Histogram.Hence The OTSU or any Global Thresholding Techniques will Fail to segment Properly.

### -->`When Images have UnEven Lightning or Illumination Conditions`:-In this case the Pixel Intensity Histogram no longer remains Bi-Modal,Hence the Thresh Value(T) calculated is not Optimal,hence resulting in Bad Segemtation Results. 

### -->When the ForeGround and BackGround doesnot have Equal Variance.

### -->When the Object area is Very Small compared to the BackGround.

### Working with Images that fall in the Limitation Categories Describe Above

In [147]:
text1 = cv2.imread(r"E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\InputData\text1.jpg")
text2 = cv2.imread(r"E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\InputData\text2.jpg")
text3 = cv2.imread(r"E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\InputData\text3.jpg")
sudoku = cv2.imread(r"E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\InputData\sudoku.png")
letters = cv2.imread(r"E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\InputData\letters.png")

In [148]:
text1_gray = cv2.cvtColor(text1,cv2.COLOR_BGR2GRAY)
text2_gray = cv2.cvtColor(text2,cv2.COLOR_BGR2GRAY)
text3_gray = cv2.cvtColor(text3,cv2.COLOR_BGR2GRAY)
sudoku_gray = cv2.cvtColor(sudoku,cv2.COLOR_BGR2GRAY)
letters_gray = cv2.cvtColor(letters,cv2.COLOR_BGR2GRAY)

In [4]:
cv2.imshow("Text1",text1_gray)
cv2.imshow("Text2",text2_gray)
cv2.imshow("Text3",text3_gray)
cv2.imshow("Suduko",sudoku_gray)
cv2.imshow("Letters",letters_gray)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [149]:
text1_th = cv2.threshold(text1_gray,0,255,cv2.THRESH_BINARY | cv2.THRESH_OTSU)
text2_th = cv2.threshold(text2_gray,0,255,cv2.THRESH_BINARY | cv2.THRESH_OTSU)
text3_th = cv2.threshold(text3_gray,0,255,cv2.THRESH_BINARY | cv2.THRESH_OTSU)
sudoku_th = cv2.threshold(sudoku_gray,0,255,cv2.THRESH_BINARY | cv2.THRESH_OTSU)
letters_th = cv2.threshold(letters_gray,0,255,cv2.THRESH_BINARY | cv2.THRESH_OTSU)

In [6]:
text1_th

(58.0,
 array([[255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        ...,
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255]], dtype=uint8))

In [7]:
text2_th

(95.0,
 array([[  0,   0,   0, ...,   0,   0,   0],
        [  0,   0,   0, ...,   0,   0,   0],
        [  0,   0,   0, ...,   0,   0,   0],
        ...,
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255]], dtype=uint8))

In [8]:
text3_th

(147.0,
 array([[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=uint8))

In [9]:
sudoku_th

(97.0,
 array([[255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        [255, 255, 255, ..., 255, 255, 255],
        ...,
        [  0,   0,   0, ..., 255, 255, 255],
        [  0,   0,   0, ..., 255, 255, 255],
        [  0,   0,   0, ..., 255, 255, 255]], dtype=uint8))

In [10]:
letters_th

(135.0,
 array([[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=uint8))

In [11]:
cv2.imshow("Text1 Thresh",text1_th[1])
cv2.imshow("Text2 Thresh",text2_th[1])
cv2.imshow("Text3 Thresh",text3_th[1])
cv2.imshow("Sudoku Thresh",sudoku_th[1])
cv2.imshow("letters thresh",letters_th[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see that in above Cells we used Both Bad Thresholding Results for all the Above Images ,Hence this Demonstrates the Limitation of OTSU and Global Thesholding Techniques. 

## ---------------------------------------------------------------------------------------------------------

### Now both Simple Thresholding and OTSU Thresholding are `Global Thresholding Technique` which means that same value of T is used to test all the Pixels in the Input Image thus segmenting BackGround and ForeGround Pixels.

### This Limitation or Problem can be overcome by using `Adaptive` or `Local Thresholding`.

## Adaptive Thresholding 

### In Adaptive Thresholding,It considers Small Neighborhood of Pixels and then finds the Optimal Threshold Value(T) for that very Small Neighborhood ,which accounts for Different Threshold Value for Different Regions thus yeilding Better Results for UnEven Lighting and Illumination Conditions.

### The General Assumption of Adaptive or Local Thresholding Technique is that Smaller Region of an Image are more likely to have approximately Uniform Illumination.

## Mathematics Of Adaptive Thresholding

### We use either use Arithmetic Mean or Guassian Mean of the Pixel Intensities in each region to Calculate the Threshold Value(T) of that Region.

### -->In Arithmetic Mean each Pixel in the Neighborhood Equally Contributes to calculate Threshold Value(T).
### -->In Gaussian Mean the Pixel far from the Center of the Neighborhood contributes Far less than Overall in calculation of Threshold Value(T).

### The General Formula of Computing Neighborhood Threshold :-
### T = mean - C

### Here C is a constant used to Fine Tune the Thresholding Value.

### We use `cv2.adaptiveThreshold()` Method to perfrom Adaptive Thresholding in OpenCV.


### For `cv2.adaptiveThreshold()` Method:-

### < First Argument >:-It takes the NUMPY Array of the Input Image on which we want to perform Adaptive Thresholding.
### < Second Argument >:-It takes the Maximum Value that will be used for THRESH Binary and THRESH Binary Inverse Techniques.
### < Third Argument >:-It takes the Flag of the ADAPTIVE Thresholding Technique we want to use.
### < Fourth Argument >:-It takes the Simple Thresholding Type we want to use.
### < Fifth Argument >:-It takes the Block Size which will be the Size of Neighborhood Region we will use in the Adaptive Thresholding Technique.
### < Sixth Argument >:-It takes the Value of C which is used to Fine Tune.

#### cv2.adaptiveThresholding():-[LINK](https://docs.opencv.org/4.x/d7/d1b/group__imgproc__misc.html#ga72b913f352e4a1b1b397736707afcde3)

#### ADAPTIVE Thresholding FLAGS :-[LINK](https://docs.opencv.org/4.x/d7/d1b/group__imgproc__misc.html#gaa42a3e6ef26247da787bf34030ed772c)

###  -->Trying Adaptive Thresh Mean :-

In [13]:
cn1gray.shape

(308, 350)

In [14]:
ad_5_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,5,1)
ad_11_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,11,1)
ad_31_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,31,1)
ad_51_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,51,1)
ad_71_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,71,1)
ad_91_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,91,1)
ad_111_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,111,1)
ad_131_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,131,1)
ad_151_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,151,1)
ad_171_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,171,1)
ad_191_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,191,1)
ad_211_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,211,1)

In [18]:
cv2.imshow("Simple OTSU Thresholding",cn1_TBO[1])
cv2.imshow("AD 5_1",ad_5_1)
cv2.imshow("AD 11_1",ad_11_1)
cv2.imshow("AD 31_1",ad_31_1)
cv2.imshow("AD 51_1",ad_51_1)
cv2.imshow("AD 71_1",ad_71_1)
cv2.imshow("AD 91_1",ad_91_1)
cv2.imshow("AD 111_1",ad_111_1)
cv2.imshow("AD 131_1",ad_131_1)
cv2.imshow("AD 151_1",ad_151_1)
cv2.imshow("AD 171_1",ad_171_1)
cv2.imshow("AD 191_1",ad_191_1)
cv2.imshow("AD 211_1",ad_211_1)
cv2.waitKey(0)
cv2.destroyAllWindows()

###  We can see in the Above Cells

### -->That Less the Block Size more the Details(ForeGround and BackGround) are segmented.
### -->On Increasing the Block Size to lose the Details while Segementation.
### --> We can see that there are an Optimal Range of Block Sizes between which we can get better Results ,After that range we tend to Lose a lot of Details in Segementation. 

In [22]:
add_271_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,271,1)
add_281_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,281,1)
add_291_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,291,1)
add_301_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,301,1)
add_311_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,311,1)
add_321_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,321,1)
add_501_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,501,1)
add_1111_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,1111,1)

In [29]:
cv2.imshow("AD 271_1",add_271_1)
cv2.imshow("AD 281_1",add_281_1)
cv2.imshow("AD 291_1",add_291_1)
cv2.imshow("AD 301_1",add_301_1)
cv2.imshow("AD 311_1",add_311_1)
cv2.imshow("AD 321_1",add_321_1)
cv2.imshow("AD 501_1",add_501_1)
cv2.imshow("AD 1111_1",add_1111_1)
cv2.waitKey(0)
cv2.destroyAllWindows()

### The Above Cells were an Experiment to See any Adnormal Behaviour when increasing the Block Size too much 

### -->It just showed that Bigger the Block Size more the Segementation will be Hurted.

In [24]:
ad_51_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,51,1)
ad_51_7 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,51,7)
ad_51_21 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,51,21)
ad_51_31 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,51,31)
ad_51_41 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,51,41)

ad_111_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,111,1)
ad_111_21 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,111,21)
ad_111_51 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,111,51)
ad_111_71 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,111,71)
ad_111_99 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,111,99)

In [25]:
cv2.imshow("AD 51_1",ad_51_1)
cv2.imshow("AD 51_7",ad_51_7)
cv2.imshow("AD 51_21",ad_51_21)
cv2.imshow("AD 51_31",ad_51_31)
cv2.imshow("AD 51_41",ad_51_41)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [26]:
cv2.imshow("AD 111_1",ad_111_1)
cv2.imshow("AD 111_21",ad_111_21)
cv2.imshow("AD 111_51",ad_111_51)
cv2.imshow("AD 111_71",ad_111_71)
cv2.imshow("AD 111_99",ad_111_99)
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see in the Above Cell 

### --->That on Increasing the C value we start to loss Boundaries and Unnecessary BackGround Noise is added in Foreground.

### --->Hence we can notice that we should use Compartibly Very Small Value of C in Relation to the Block Size value.

### Example for:-
#### ---->For 51 Block Size C can be between 1-7 range.
#### ---->For 111 Block Size C can be betweeen 5-21 range.

In [27]:
ad_51_51 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,51,51)
ad_111_111 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,111,111)
ad_51_91 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,51,91) 

In [28]:
cv2.imshow("AD 51_51",ad_51_51)
cv2.imshow("AD 111_111",ad_111_111)
cv2.imshow("AD 51_91",ad_51_91)
cv2.waitKey(0)
cv2.destroyAllWindows()

### We tried Odd Combination Value of Block Size and  C and got Bad Results

### -->Trying Adaptive Thresh Gaussian 

In [30]:
adg_5_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,5,1)
adg_11_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,11,1)
adg_31_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,31,1)
adg_51_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,51,1)
adg_71_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,71,1)
adg_91_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,91,1)
adg_111_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,111,1)
adg_131_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,131,1)
adg_151_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,151,1)
adg_171_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,171,1)
adg_191_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,191,1)
adg_211_1 = cv2.adaptiveThreshold(cn1gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,211,1)

In [35]:
cv2.imshow("Simple OTSU Thresholding",cn1_TBO[1])
cv2.imshow("ADG 5_1",adg_5_1)
cv2.imshow("ADG 11_1",adg_11_1)
cv2.imshow("ADG 31_1",adg_31_1)
cv2.imshow("ADG 51_1",adg_51_1)
cv2.imshow("ADG 71_1",adg_71_1)
cv2.imshow("ADG 91_1",adg_91_1)
cv2.imshow("ADG 111_1",adg_111_1)
cv2.imshow("ADG 131_1",adg_131_1)
cv2.imshow("ADG 151_1",adg_151_1)
cv2.imshow("ADG 171_1",adg_171_1)
cv2.imshow("ADG 191_1",adg_191_1)
cv2.imshow("ADG 211_1",adg_211_1)
cv2.waitKey(0)
cv2.destroyAllWindows()

###  We can see in the Above Cells

### -->That Less the Block Size more the Details(ForeGround and BackGround) are segmented.
### -->On Increasing the Block Size to lose the Details while Segementation.
### --> We can see that there are an Optimal Range of Block Sizes between which we can get better Results ,After that range we tend to Lose a lot of Details in Segementation. 

In [37]:
cv2.imshow("AD_51_1",ad_51_1)
cv2.imshow("ADG_51_1",adg_51_1)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [38]:
cv2.imshow("AD_91_1",ad_91_1)
cv2.imshow("ADG_91_1",adg_91_1)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [39]:
cv2.imshow("AD_151_1",ad_151_1)
cv2.imshow("ADG_151_1",adg_151_1)
cv2.waitKey(0)
cv2.destroyAllWindows()

### In the Above Cells we compared the Results of the Adaptive Gaussian and Mean (with same Block Size and C Values) and we cann notice that:-

### -->Adaptive Gaussian by its defination Segments the Smaller Regions in better ways and is also reluctant to Noise.

### -->Now Trying Adaptive Thresholding on the Images where OTSU Failed to perform Better

### Trying for text2 Image:-

In [45]:
text2.shape

(300, 300, 3)

In [150]:
text2_adt_21_1 = cv2.adaptiveThreshold(text2_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,21,1)
text2_adt_11_1 = cv2.adaptiveThreshold(text2_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,11,1)
text2_adt_31_1 = cv2.adaptiveThreshold(text2_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,31,1)
text2_adt_51_1 = cv2.adaptiveThreshold(text2_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,51,1)
text2_adt_43_1 = cv2.adaptiveThreshold(text2_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,43,1)
text2_adt_37_9 = cv2.adaptiveThreshold(text2_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,37,9)
text2_adt_35_5 = cv2.adaptiveThreshold(text2_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,35,5)
text2_adt_55_21 = cv2.adaptiveThreshold(text2_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,55,21)
text2_adt_55_11 = cv2.adaptiveThreshold(text2_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,55,11)

In [152]:
cv2.imshow("Gray Scale Image",text2_gray)
cv2.imshow("OTSU",text2_th[1])
cv2.imshow("21 1",text2_adt_21_1)
cv2.imshow("11 1",text2_adt_11_1)
cv2.imshow("31 1",text2_adt_31_1)
cv2.imshow("51 1",text2_adt_51_1)
cv2.imshow("43 1",text2_adt_43_1)
cv2.imshow("37 9",text2_adt_37_9)
cv2.imshow("35 5",text2_adt_35_5)
cv2.imshow("55 25",text2_adt_55_21)
cv2.imshow("55 11",text2_adt_55_11)
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see after Trying with Various Block Size and C values we can see we got the Best Results with 37_9.

### --->But we can also Notice that there are some values around Specfic Range which can give us the Good Results

In [77]:
text3.shape

(691, 1000, 3)

In [156]:
text3_ad_61_11 = cv2.adaptiveThreshold(text3_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,61,11)
text3_ad_75_25 = cv2.adaptiveThreshold(text3_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,75,25)
text3_ad_75_11 = cv2.adaptiveThreshold(text3_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,75,11)

In [157]:
cv2.namedWindow("Normal GrayScale Image",cv2.WINDOW_NORMAL)
cv2.namedWindow("61 11",cv2.WINDOW_NORMAL)
cv2.namedWindow("75 25",cv2.WINDOW_NORMAL)
cv2.namedWindow("75 11",cv2.WINDOW_NORMAL)
cv2.imshow("Normal GrayScale Image",text3_gray)
cv2.imshow("OTSU Thresholding",text3_th[1])
cv2.imshow("61 11",text3_ad_61_11)
cv2.imshow("75 11",text3_ad_75_11)
cv2.imshow("75 25",text3_ad_75_25)
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see after Trying with Various Block Size and C values we can see we got the Best Results with 75_11.

### --->But we can also Notice that there are some values around Specfic Range which can give us the Good Results

### --->And we can also Notice that High C Values tends to Bad Results. 

In [87]:
sudoku.shape

(563, 558, 3)

In [158]:
sudoku_ad_47_9 = cv2.adaptiveThreshold(sudoku_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,47,9)
sudoku_ad_11_9 = cv2.adaptiveThreshold(sudoku_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,11,1)
sudoku_ad_31_13 = cv2.adaptiveThreshold(sudoku_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,31,13)
sudoku_ad_47_11 = cv2.adaptiveThreshold(sudoku_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,47,9)

In [159]:
cv2.imshow("Gray Scale",sudoku_gray)
cv2.imshow("OTSU Thresholding",sudoku_th[1])
cv2.imshow("47 9",sudoku_ad_47_9)
cv2.imshow("11 9",sudoku_ad_11_9)
cv2.imshow("31 13",sudoku_ad_31_13)
cv2.imshow("47 11",sudoku_ad_47_11)
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see after Trying with Various Block Size and C values we can see we got the Best Results with 47_11.

### --->But we can also Notice that there are some values around Specfic Range which can give us the Good Results

In [104]:
letters.shape

(220, 318, 3)

In [160]:
letters_ad_11_1 = cv2.adaptiveThreshold(letters_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,11,1)
letters_ad_21_1 = cv2.adaptiveThreshold(letters_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,21,1)
letters_ad_27_5 = cv2.adaptiveThreshold(letters_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,27,5)
letters_ad_33_1 = cv2.adaptiveThreshold(letters_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,31,1)
letters_ad_33_5 = cv2.adaptiveThreshold(letters_gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,31,5)

In [161]:
cv2.imshow("GrayScale Image",letters_gray)
cv2.imshow("OTSU Threshold",letters_th[1])
cv2.imshow("11 1",letters_ad_11_1)
cv2.imshow("21 1",letters_ad_21_1)
cv2.imshow("27 5",letters_ad_27_5)
cv2.imshow("33 1",letters_ad_33_1)
cv2.imshow("33 5",letters_ad_33_5)
cv2.waitKey(0)
cv2.destroyAllWindows()

### We can see after Trying with Various Block Size and C values we can see we got the Best Results with 33_5.

### --->But we can also Notice that there are some values around Specfic Range which can give us the Good Results

### Hence we can notice that The Block Size Right Value can be estimated through the Image Size Relation

### **NOTE:-There is a Range of Block Size and C Value in which we can get the Best Results.

## --------------------------------------------------------------------------------------------------------------------------------